### PR TITLE
Dmatrix refactor stage 2

### PR DESCRIPTION
**Pull request by RAMitchell**
- Remove buffered rowset usage where possible
- Transition to c++11 style iterators for row access

To do:
- Transition column iterators to C++ 11
- Consider removal of DataSource class. SimpleCSRSource and SparsePageSource are so tightly coupled to their respective dmatrix types that it makes no sense for them to be separate classes. Removing these classes would significantly reduce the complexity of dmatrix.

### CODE REVIEW

**Comment by RAMitchell** [comment]
There is a major change in behaviour here. External memory column pages are no longer generated, they are constructed on the fly from row pages. I still need to do some basic timing experiments to check performance has not degraded.

**Comment by RAMitchell** [comment]
My idea to no longer generate column pages was a failure, performance degraded significantly from having to transpose and sort the row pages everytime on the fly. I brought back column pages except this time I reused all of the code within sparse page source to manage them. This removed a huge amount of duplication that existed with row pages and column pages.

Here are some performance numbers:
```
Higgs 100k instances
Master:
[99]	train-error:0.21775
External memory time: 13.425671815872192
[99]	train-error:0.20488
Internal memory time: 4.167760610580444

DMatrix refactor:
[99]	train-error:0.21464
External memory time: 11.791306734085083
[99]	train-error:0.20488
Internal memory time: 4.531860113143921

Higgs 1M instances:
Master:
[99]	train-error:0.261237
External memory time: 75.71507811546326
[99]	train-error:0.251887
Internal memory time: 54.69732618331909
DMatrix refactor:
[99]	train-error:0.262015
External memory time: 69.23167657852173
[99]	train-error:0.251887
Internal memory time: 57.146806955337524
```

@hcho3 can I please get a review?

File: src/linear/coordinate_common.h [hunk 1]
```diff
@@ -126,12 +122,11 @@ inline std::pair<double, double> GetGradientParallel(int group_idx, int num_grou
 inline std::pair<double, double> GetBiasGradientParallel(int group_idx, int num_group,
                                                          const std::vector<GradientPair> &gpair,
                                                          DMatrix *p_fmat) {
-  const RowSet &rowset = p_fmat->BufferedRowset();
```
Comment (hcho3) [inline src/linear/coordinate_common.h]: Can you provide your reasoning behind removing `BufferedRowset`? Is it the case that DMatrix always stores every row in `buffered_rowset_`?

File: src/data/sparse_page_dmatrix.cc [hunk 2]
```diff
@@ -12,261 +12,91 @@
...
+  return BatchSet(begin_iter);
 }
 
-void SparsePageDMatrix::InitColAccess(
-  size_t max_row_perbatch, bool sorted) {
-  if (HaveColAccess(sorted)) return;
-  if (TryInitColData(sorted)) return;
-  const MetaInfo& info = this->Info();
-  if (max_row_perbatch == std::numeric_limits<size_t>::max()) {
-    max_row_perbatch = kMaxRowPerBatch;
```
Comment (hcho3) [inline src/data/sparse_page_dmatrix.cc]: I notice that there is no reference to `max_row_perbatch` or `kMaxRowPerBatch` anywhere in your codebase. Are we exclusively using `kPageSize` to determine page sizes? If so, we should remove `max_row_perbatch` parameter and `kMaxRowPerBatch`  constant.

File: src/tree/updater_colmaker.cc [hunk 3]
```diff
@@ -254,13 +246,13 @@ class ColMaker: public TreeUpdater {
     // this function does not support nested functions
     inline void ParallelFindSplit(const SparsePage::Inst &col,
                                   bst_uint fid,
-                                  const DMatrix &fmat,
```
Comment (hcho3) [inline src/tree/updater_colmaker.cc]: Why the change from `const DMatrix &fmat` to `DMatrix *p_fmat` here?

File: src/linear/coordinate_common.h (hunk 1 above)
Comment (RAMitchell) [reply src/linear/coordinate_common.h]: Yes every row is always used after we removed prob_buf_row.

File: src/tree/updater_colmaker.cc (hunk 3 above)
Comment (RAMitchell) [reply src/tree/updater_colmaker.cc]: The function p_fmat->GetColDensity() lazily generates and stores the column densities on demand, therefore on the first call this modifies data in dmatrix.

File: src/data/sparse_page_dmatrix.cc (hunk 2 above)
Comment (RAMitchell) [reply src/data/sparse_page_dmatrix.cc]: Yes, I will do this.

**Comment by hcho3** [comment]
@RAMitchell I'm going off to a 3-day vacation. Will come back to this afterwards.

**Comment by hcho3** [comment]
@RAMitchell  Please ignore the failed test `continuous-integration/jenkins/pr-head`. (This was due to misconfiguration of Jenkins CI server.) Only `continuous-integration/jenkins/pr-merge` is relevant. I will give you additional feedback in a day or two.

File: tests/cpp/data/test_simple_dmatrix.cc [hunk 4]
```diff
@@ -46,40 +45,18 @@ TEST(SimpleDMatrix, ColAccessWithoutBatches) {
...
   // Loop over the batches and assert the data is as expected
   long num_col_batch = 0;
-  col_iter->BeforeFirst();
-  while (col_iter->Next()) {
+  for (const auto &batch : dmat->GetSortedColumnBatches()) {
     num_col_batch += 1;
-    EXPECT_EQ(col_iter->Value().Size(), dmat->Info().num_col_)
-      << "Expected batch size = number of cells as #batches is 1.";
-    for (int i = 0; i < static_cast<int>(col_iter->Value().Size()); ++i) {
-      EXPECT_EQ(col_iter->Value()[i].size(), dmat->GetColSize(i))
```
Comment (hcho3) [inline tests/cpp/data/test_simple_dmatrix.cc]: I feel that this check should be replaced, not just removed. Is this test redundant?

File: tests/cpp/data/test_sparse_page_dmatrix.cc [hunk 5]
```diff
@@ -19,33 +18,34 @@ TEST(SparsePageDMatrix, MetaInfo) {
...
+  // Create sufficiently large data to make two row pages
+  std::string tmp_file = CreateBigTestData(5000000);
   xgboost::DMatrix * dmat = xgboost::DMatrix::Load(
     tmp_file + "#" + tmp_file + ".cache", true, false);
   std::remove(tmp_file.c_str());
   EXPECT_TRUE(FileExists(tmp_file + ".cache.row.page"));
 
-  auto row_iter = dmat->RowIterator();
-  // Loop over the batches and count the records
+   //Loop over the batches and count the records
```
Comment (hcho3) [inline tests/cpp/data/test_sparse_page_dmatrix.cc]: Space: `// Loop over the batches and count the records`

File: tests/cpp/data/test_sparse_page_dmatrix.cc [hunk 6]
```diff
@@ -19,33 +18,34 @@ TEST(SparsePageDMatrix, MetaInfo) {
...
   long row_count = 0;
-  row_iter->BeforeFirst();
-  while (row_iter->Next()) row_count += row_iter->Value().Size();
+  for(auto &batch:dmat->GetRowBatches())
```
Comment (hcho3) [inline tests/cpp/data/test_sparse_page_dmatrix.cc]: `for (auto& batch : dmat->GetRowBatches()) {`

File: tests/cpp/data/test_sparse_page_dmatrix.cc [hunk 7]
```diff
@@ -19,33 +18,34 @@ TEST(SparsePageDMatrix, MetaInfo) {
...
+  {
+    row_count += batch.Size();
+  }
   EXPECT_EQ(row_count, dmat->Info().num_row_);
   // Test the data read into the first row
-  row_iter->BeforeFirst();
-  row_iter->Next();
-  auto first_row = row_iter->Value()[0];
+  auto &batch = *dmat->GetRowBatches().begin();
```
Comment (hcho3) [inline tests/cpp/data/test_sparse_page_dmatrix.cc]: Add a test for non-emptiness of `dmat->GetRowBatches()`?

File: tests/cpp/data/test_sparse_page_dmatrix.cc [hunk 8]
```diff
@@ -59,34 +59,25 @@ TEST(SparsePageDMatrix, ColAccess) {
   xgboost::DMatrix * dmat = xgboost::DMatrix::Load(
     tmp_file + "#" + tmp_file + ".cache", true, false);
   std::remove(tmp_file.c_str());
-  EXPECT_FALSE(FileExists(tmp_file + ".cache.col.page"));
```
Comment (hcho3) [inline tests/cpp/data/test_sparse_page_dmatrix.cc]: We may want to test the existence of `col.page` file. Or does `col.page` file not generated at all?

File: tests/cpp/data/test_sparse_page_dmatrix.cc [hunk 9]
```diff
@@ -59,34 +59,25 @@ TEST(SparsePageDMatrix, ColAccess) {
...
+    EXPECT_EQ(col_batch.Size(), dmat->Info().num_col_);
+    EXPECT_EQ(col_batch[1][0].fvalue, 10.0f);
+    EXPECT_EQ(col_batch[1].size(), 1);
   }
-  EXPECT_EQ(num_col_batch, dmat->Info().num_row_)
-    << "Expected num batches to be same as num_rows as max_row_perbatch is 1";
-  col_iter = nullptr;
 
   std::remove((tmp_file + ".cache").c_str());
-  std::remove((tmp_file + ".cache.col.page").c_str());
```
Comment (hcho3) [inline tests/cpp/data/test_sparse_page_dmatrix.cc]: Is `col.page` file not generated at all? Why? Also how about `sorted.col.page file?

File: tests/cpp/tree/test_gpu_hist.cu [hunk 10]
```diff
@@ -23,14 +23,10 @@ TEST(gpu_hist_experimental, TestSparseShard) {
...
 
-  dmlc::DataIter<SparsePage>* iter = (*dmat)->RowIterator();
-  iter->BeforeFirst();
-  CHECK(iter->Next());
-  const SparsePage& batch = iter->Value();
+  const SparsePage& batch = *(*dmat)->GetRowBatches().begin();
   DeviceShard shard(0, 0, 0, rows, p);
   shard.InitRowPtrs(batch);
   shard.InitCompressedData(gmat.cut, batch);
-  CHECK(!iter->Next());
```
Comment (hcho3) [inline tests/cpp/tree/test_gpu_hist.cu]: Why remove this check?

File: tests/cpp/tree/test_gpu_hist.cu [hunk 11]
```diff
@@ -64,15 +60,10 @@ TEST(gpu_hist_experimental, TestDenseShard) {
...
-  dmlc::DataIter<SparsePage>* iter = (*dmat)->RowIterator();
-  iter->BeforeFirst();
-  CHECK(iter->Next());
-  const SparsePage& batch = iter->Value();
-
+  const SparsePage& batch = *(*dmat)->GetRowBatches().begin();
   DeviceShard shard(0, 0, 0, rows, p);
   shard.InitRowPtrs(batch);
   shard.InitCompressedData(gmat.cut, batch);
-  CHECK(!iter->Next());
```
Comment (hcho3) [inline tests/cpp/tree/test_gpu_hist.cu]: Why remove this check?

**Comment by hcho3** [comment]
And it looks like the latest commit in `master` introduced a conflict. Can you look at it?

File: src/data/simple_dmatrix.cc [hunk 12]
```diff
@@ -4,103 +4,76 @@
...
-      }
-    }
-  }
+ private:
+  SparsePage* page_{nullptr};
+};
 
-  CHECK_EQ(pcol->Size(), Info().num_col_);
+BatchSet SimpleDMatrix::GetRowBatches() {
+  auto cast = dynamic_cast<SimpleCSRSource*>(source_.get());
```
[the remark this answers is not in the document]
Comment (RAMitchell) [reply src/data/simple_dmatrix.cc]: The unique pointer is passed to the constructor like this:
`
explicit SimpleDMatrix(std::unique_ptr<DataSource>&& source)
`

It's not obvious to me how to safely cast this type. I would prefer to defer this change until we get rid of the DataSource class.

File: include/xgboost/data.h [hunk 13]
```diff
@@ -251,6 +296,47 @@ class SparsePage {
...
+  virtual ~BatchIteratorImpl() {}
+  virtual BatchIteratorImpl* Clone() = 0;
+  virtual const SparsePage& operator*() const = 0;
+  virtual void operator++() = 0;
+  virtual bool AtEnd() const = 0;
+};
+
+class BatchIterator {
+ public:
+  using iterator_category = std::forward_iterator_tag;
```
[the remark this answers is not in the document]
Comment (RAMitchell) [reply include/xgboost/data.h]: It is used by the BatchSet class.

File: tests/cpp/data/test_sparse_page_dmatrix.cc (hunk 7 above)
Comment (RAMitchell) [reply tests/cpp/data/test_sparse_page_dmatrix.cc]: I think emptiness will be caught by the existing tests.

File: tests/cpp/data/test_sparse_page_dmatrix.cc (hunk 8 above)
Comment (RAMitchell) [reply tests/cpp/data/test_sparse_page_dmatrix.cc]: Done.

File: tests/cpp/data/test_sparse_page_dmatrix.cc (hunk 9 above)
Comment (RAMitchell) [reply tests/cpp/data/test_sparse_page_dmatrix.cc]: Done.

File: tests/cpp/tree/test_gpu_hist.cu (hunk 10 above)
Comment (RAMitchell) [reply tests/cpp/tree/test_gpu_hist.cu]: This just checks and there is only a single batch. I think this is unnecessary for this test, it would not matter if the dmatrix had multiple batches.

File: src/data/sparse_page_dmatrix.cc [hunk 14]
```diff
@@ -12,261 +12,91 @@
...
-        } else {
-          return fmt->Read(*dptr, fi, index_set_);
-        }
-      },  [this, fi, fbegin] () {
-        fi->Seek(fbegin);
-        index_set_ = set_index_set_;
-        load_all_ = set_load_all_;
-      });
-  }
-}
```
[the remark this answers is not in the document]
Comment (RAMitchell) [reply src/data/sparse_page_dmatrix.cc]: The column pages are now managed by SparsePageSource, so it does in fact use prefetcher_ when iterating through column pages.

File: src/data/sparse_page_dmatrix.cc [hunk 15]
```diff
@@ -12,261 +12,91 @@
...
+BatchSet SparsePageDMatrix::GetRowBatches() {
+  auto cast = dynamic_cast<SparsePageSource*>(row_source_.get());
+  cast->BeforeFirst();
+  cast->Next();
+  auto begin_iter = BatchIterator(new SparseBatchIteratorImpl(cast));
+  return BatchSet(begin_iter);
 }
 
-bool SparsePageDMatrix::TryInitColData(bool sorted) {
-  // load meta data.
```
[the remark this answers is not in the document]
Comment (RAMitchell) [reply src/data/sparse_page_dmatrix.cc]: Yes, I have effectively consolidated two separate blocks of code for managing pages into one, using only SparsePageSource.

File: src/data/sparse_page_dmatrix.cc [hunk 16]
```diff
@@ -12,261 +12,91 @@
...
-  return col_iter_.get();
+ private:
+  SparsePageSource* source_{nullptr};
+  bool at_end_{ false };
+};
+
+BatchSet SparsePageDMatrix::GetRowBatches() {
+  auto cast = dynamic_cast<SparsePageSource*>(row_source_.get());
+  cast->BeforeFirst();
+  cast->Next();
```
[the remark this answers is not in the document]
Comment (RAMitchell) [reply src/data/sparse_page_dmatrix.cc]: I am using the DataSource class to do the actual iteration, these two lines ensure it is reset to the first row batch, which otherwise may not be true.

File: src/data/sparse_page_dmatrix.cc [hunk 17]
```diff
@@ -12,261 +12,91 @@
...
+  auto begin_iter = BatchIterator(new SparseBatchIteratorImpl(cast));
+  return BatchSet(begin_iter);
 }
 
-bool SparsePageDMatrix::TryInitColData(bool sorted) {
-  // load meta data.
-  std::vector<std::string> cache_shards = common::Split(cache_info_, ':');
-  {
-    std::string col_meta_name = cache_shards[0] + ".col.meta";
```
[the remark this answers is not in the document]
Comment (RAMitchell) [reply src/data/sparse_page_dmatrix.cc]: buffered_rowset_ & &col_size_ do not exist any more, making this redundant.

File: src/data/sparse_page_dmatrix.cc [hunk 18]
```diff
@@ -12,261 +12,91 @@
...
-    std::unique_ptr<dmlc::SeekStream> fdata(
-        dmlc::SeekStream::CreateForRead(col_data_name.c_str(), true));
-    if (fdata == nullptr) return false;
-    files.push_back(std::move(fdata));
+BatchSet SparsePageDMatrix::GetSortedColumnBatches() {
+  // Lazily instantiate
+  if (!sorted_column_source_) {
+    SparsePageSource::CreateColumnPage(this, cache_info_, true);
+    sorted_column_source_.reset(
+        new SparsePageSource(cache_info_, ".sorted.col.page"));
```
[the remark this answers is not in the document]
Comment (RAMitchell) [reply src/data/sparse_page_dmatrix.cc]: Done.

File: src/tree/updater_gpu_hist.cu [hunk 19]
```diff
@@ -803,10 +803,8 @@ class GPUHistMaker : public TreeUpdater {
     std::vector<size_t> row_segments;
     dh::RowSegments(info_->num_row_, n_devices, &row_segments);
 
-    dmlc::DataIter<SparsePage>* iter = dmat->RowIterator();
-    iter->BeforeFirst();
-    CHECK(iter->Next()) << "Empty batches are not supported";
-    const SparsePage& batch = iter->Value();
+    auto batch_iter = dmat->GetRowBatches().begin();
+    const SparsePage& batch = *batch_iter;
```
[the remark this answers is not in the document]
Comment (RAMitchell) [reply src/tree/updater_gpu_hist.cu]: I have added null checks to the iterators.

File: tests/cpp/data/test_simple_dmatrix.cc (hunk 4 above)
Comment (RAMitchell) [reply tests/cpp/data/test_simple_dmatrix.cc]: Replaced with an equivalent test.
